code/modules/holodeck/computer.dm

```diff
@@ -110,207 +126,267 @@
+				do_sparks(2, 1, holo_turf)
+			SSexplosions.lowturf += holo_turf
+			holo_turf.hotspot_expose(1000,500,1)
 
 	if(!(obj_flags & EMAGGED))
 		for(var/item in spawned)
 			if(!(get_turf(item) in linked))
-				derez(item, 0)
-	for(var/e in effects)
-		var/obj/effect/holodeck_effect/HE = e
-		HE.tick()
-
+				derez(item)
+	for(var/_effect in effects)
+		var/obj/effect/holodeck_effect/holo_effect = _effect
+		holo_effect.tick()
 	active_power_usage = 50 + spawned.len * 3 + effects.len * 5
 
-/obj/machinery/computer/holodeck/emag_act(mob/user)
-	if(obj_flags & EMAGGED)
-		return
-	if(!LAZYLEN(emag_programs))
-		to_chat(user, "[src] does not seem to have a card swipe port. It must be an inferior model.")
-		return
-	playsound(src, "sparks", 75, TRUE)
-	obj_flags |= EMAGGED
-	to_chat(user, "<span class='warning'>You vastly increase projector power and override the safety and security protocols.</span>")
-	say("Warning. Automatic shutoff and derezzing protocols have been corrupted. Please call Nanotrasen maintenance and do not use the simulator.")
-	log_game("[key_name(user)] emagged the Holodeck Control Console")
-	nerf(!(obj_flags & EMAGGED))
-
-/obj/machinery/computer/holodeck/emp_act(severity)
-	. = ..()
-	if(. & EMP_PROTECT_SELF)
-		return
-	emergency_shutdown()
-
-/obj/machinery/computer/holodeck/ex_act(severity, target)
-	emergency_shutdown()
-	return ..()
-
-/obj/machinery/computer/holodeck/blob_act(obj/structure/blob/B)
-	emergency_shutdown()
-	return ..()
-
-/obj/machinery/computer/holodeck/proc/generate_program_list()
-	for(var/typekey in subtypesof(program_type))
-		var/area/holodeck/A = GLOB.areas_by_type[typekey]
-		if(!A || !A.contents.len)
-			continue
-		var/list/info_this = list()
-		info_this["name"] = A.name
-		info_this["type"] = A.type
-		if(A.restricted)
-			LAZYADD(emag_programs, list(info_this))
-		else
-			LAZYADD(program_cache, list(info_this))
-
 /obj/machinery/computer/holodeck/proc/toggle_power(toggleOn = FALSE)
 	if(active == toggleOn)
 		return
 
 	if(toggleOn)
-		if(last_program && last_program != offline_program)
-			addtimer(CALLBACK(src, .proc/load_program, last_program, TRUE), 25)
+		if(last_program && (last_program != offline_program))
+			addtimer(CALLBACK(src,.proc/load_program, last_program, TRUE), 25)
 		active = TRUE
 	else
 		last_program = program
 		load_program(offline_program, TRUE)
 		active = FALSE
 
+/obj/machinery/computer/holodeck/power_change()
+	. = ..()
+	INVOKE_ASYNC(src, .proc/toggle_power, !machine_stat)
+
+///shuts down the holodeck and force loads the offline_program
 /obj/machinery/computer/holodeck/proc/emergency_shutdown()
 	last_program = program
-	load_program(offline_program, TRUE)
 	active = FALSE
+	load_program(offline_program, TRUE)
 
+///returns TRUE if the entire floor of the holodeck is intact, returns FALSE if any are broken
 /obj/machinery/computer/holodeck/proc/floorcheck()
-	for(var/turf/T in linked)
-		if(!T.intact || isspaceturf(T))
+	for(var/turf/holo_floor in linked)
```

Comment:
Is anything that's not a turf inside `linked` ?

Reply:
yes every atom that isnt an area inside of the holodeck room where the programs are loaded into is inside of linked